File: ruby/udemy_class/arrays/anagrams.rb
# efficient solution

def anagrams(string1, string2)

  string_count = Hash.new { |h, k| h[k] = 0 }

  string1.split('').each do |el|
    if el != " "
      string_count[el.downcase] += 1
    end
  end

  string2.split('').each do |el|
    if el != " "
      string_count[el.downcase] -= 1
    end
  end

  string_count.each do |k, v|
    return false unless v == 0
  end

  return true
end

p anagrams("dog", "god ")
p anagrams("poo", "head")

def sort_anagrams(string1, string2)
  stripped1 = string1.gsub(' ', '')
  stripped2 = string2.gsub(' ', '')

  return stripped1.split('').sort == stripped2.split('').sort
end

p sort_anagrams("dog", "god ")
p sort_anagrams("poo", "head")

File: python/udemy_class/arrays/anagrams.py
def anagrams(string1, string2):

    character_count = {}

    for char in string1:
        char = char.lower()
        if char != ' ':
            if char in character_count:
                character_count[char] += 1
            else:
                character_count[char] = 1

    for char in string2:
        char = char.lower()
        if char != ' ':
            if char in character_count:
                character_count[char] -= 1
            else:
                character_count[char] = 1

    for k in character_count:
        if character_count[k] != 0:
            return False

    return True

print(anagrams("dog", "God"))
print(anagrams("poo", "Head"))
